dts/bindings/memory-controllers/atmel,sam-smc.yaml

```diff
@@ -0,0 +1,130 @@
+# Copyright (c) 2022, Basalte bv
+# SPDX-License-Identifier: Apache-2.0
+
+description: |
+  Atmel Static Memory Controller (SMC).
+
+  The SMC allows to interface with static-memory mapped external devices such as
+  SRAM, PSRAM, PROM, EPROM, EEPROM, LCD Module, NOR Flash and NAND Flash.
+
+  The SMC controller can have up to 4 children defining the connected external
+  memory devices. Device Tree example:
+
+  &smc {
+    status = "okay";
+    pinctrl-0 = <&smc_default>;
+    pinctrl-names = "default";
+
+    is66wv51216dbll@0 {
+      reg = <0>;
+
+      atmel,smc-write-mode = "nwe";
+      atmel,smc-read-mode = "nrd";
+      atmel,smc-setup-timing = <1 1 1 1>;
+      atmel,smc-pulse-timing = <6 6 6 6>;
+      atmel,smc-cycle-timing = <7 7>;
+    };
+  };
+
+  Finally, in order to make the memory available you will need to define new
+  memory device/s in DeviceTree:
+
+  sram1: sram@60000000 {
+      compatible = "zephyr,memory-region", "mmio-sram";
+      device_type = "memory";
+      reg = <0x60000000 DT_SIZE_K(512)>;
+      zephyr,memory-region = "SRAM1";
+  };
+
+compatible: "atmel,sam-smc"
+
+include: [base.yaml, pinctrl-device.yaml]
+
+properties:
+  reg:
+    required: true
+
+  "#address-cells":
+    required: true
+    const: 1
+
+  "#size-cells":
+    required: true
+    const: 0
+
+  peripheral-id:
+    type: int
+    description: peripheral ID
+    required: true
+
+child-binding:
+  description: |
+    Child device nodes are representing devices connected to the EBI/SMC bus.
+
+  properties:
+    reg:
+      type: int
+      required: true
+      description: |
+        The device's SMC CS number.
+        Valid range: 0 - 3
+
+    atmel,smc-write-mode:
+      type: string
+      required: false
+      default: "nwe"
+      description: |
+        Selects the mode that indicates which signal is used for the write
+        operation. Reset and default value is nwe.
+      enum:
+        - "ncs"
+        - "nwe"
+
+    atmel,smc-read-mode:
+      type: string
+      required: false
+      default: "nrd"
+      description: |
+        Selects the mode that indicates which signal is used for the read
+        operation. Reset and default value is nrd.
+      enum:
+        - "ncs"
+        - "nrd"
+
+    atmel,smc-setup-timing:
+      type: array
+      required: false
+      default: [1, 1, 1, 1]
+      description: |
+        SMC setup timing configuration in cycles. [NWE, NCS_WR, NRD, NCS_RD]
+        1 cycle is the default for devices that perform actions on rising
+        edges. Null delay setups are permitted, for consecutive reads/writes.
```

Comment:
This is kind of confusing still. It says 1 is the default but the default is actually an array. I guess the description is meant to say the first element is NWE setup timing, the second is NCS_WR, etc.? I personally find that a bit confusing if so.

My bigger question here though is why are you assuming devices that perform actions on rising edges? Would a different configuration only appear under very unusual conditions? If it's common for devices not to perform actions on rising edges, then these should be made required instead.

Reply:
I was trying to translate the datasheet a bit into the description and I thought it would be safer to have a minimal setup timing as the default for all read/write modes.
<img width="563" alt="image" src="https://user-images.githubusercontent.com/3170234/181699730-8e09f6ee-d145-4ef0-9559-bc2dd7df9664.png">

Maybe our atmel/michrochip maintainer @nandojve can suggest how the (timing) properties should be set?

Sorry this has taken already so many iterations, but I agree it has to be clear for everyone.

Reply:
Hi @pdgendt ,

The devicetree properties must be well documented. A maintainer from devicetree should read the documentation and have a very clear picture about it without inspect code or even datasheet. Otherwise no one can provide fixes in some situations and it is even more important when the property have a **default** value. I think both @mbolivar-nordic and @galak are arguing in that direction to make sure that properties are well defined and well documented.

In regards of timing parameters, I'm OK to have it as a array. I think what @mbolivar-nordic is asking to you go in the direction of have 100% sure that a  `default` is fine for this driver since the documented is not 100% clear. The default should be used only when you have a generic case that can full fill majority of scenarios otherwise `required` field should be added and set as `true`. This imposes that each board must explicit define the values since no `default` property will be available. The other think that I noted is that for someone that is not aware about SoC it is difficult to understand the default values.

If you look at my current PR #48377 you will noted that I'm in the same situation. The amount of information that I added at first push wasn't enough to make it clear. A second version was send with substantial more documentation and examples. This kind of situation happen because, in your case, the driver is one of the first and maintainers will look in deep to make sure it is clear how properties were define. Remember that people inspect code to pick ideas for drivers and this maybe be a reference for others.

Questions that may help to evaluate the properties:

* What happen if user selects another SDRAM manufacturer? Are the parameters same?
* What happen if user selects a slower speed memory? The current default configuration is set up for the slowest case?
* The current default values may work for all memories or at least for the majority of that?
* If I set up SAM4E-XPRO board will be possible to use same SDRAM parameters from SAM4S-Xplained (since memories are different)?

Sugestions:

 * Increase the amount of documentation
 * Add the sam4s_xplained configuration as an example in the devicetree documentation
 * Explain how to configure each parameter and why you select that value since it have a relationship with memory used and the SoC controller.

Reply:
Thanks @nandojve for the elaborate response.

Based on the feedback, it is clear that default values do not make sense, due to the amount of different external controllers possible.
All properties are now required without defaults.

Is the `sam4s_xplained` example configuration in the devicetree documentation sufficient?